fawltydeps/packages.py

```diff
@@ -194,8 +198,38 @@ def lookup_package(self, package_name: str) -> Optional[Package]:
         return self.packages.get(Package.normalize_name(package_name))
 
 
+def setup_venv(dep_names: Iterable[str]) -> Optional[Path]:
+    """Create the venv dir and install requirements."""
+
+    def venv_script(venv_dir: Path, dep_names: Iterable[str]) -> List[str]:
+        return (
+            [
+                f"rm -rf {venv_dir}",
+                f"python3 -m venv {venv_dir}",
+            ]
+            + [f"{venv_dir}/bin/pip install {shlex.quote(req)}" for req in dep_names]
+            + [
+                f"touch {venv_dir}/.installed",
+            ]
+        )
+
+    venv_dir = Path(tempfile.mkdtemp())
+    script = venv_script(venv_dir, dep_names)
+    subprocess.run(
+        " && ".join(script),
+        check=True,  # fail if any of the commands fail
+        shell=True,  # pass multiple shell commands to the subprocess
+    )
+    # Make sure the venv has been installed
+    assert Path(venv_dir, ".installed").is_file()
+
+    return venv_dir
```

Comment:
This is more or less copied from our tests, I assume. I'm not crazy about the amount of shell commands we end up running on the user's system here. It might be hard to avoid them all, but doing things like `rm -rf` and `touch` in a subprocess makes me cringe a little.

It made sense in the tests, because (a) tests are not run by our regular users, and (b) for the cached venvs we want a checksum that includes the operations used to create the venv. Here, I would much rather use something like

```python
import venv
...
venv.create(venv_dir, clear=True, with_pip=True)
for req in dep_names:
    subprocess.run(
        [f"{venv_dir}/bin/pip", "install", req],  # no need to shlex.quote when we don't pass through the shell
        shell=False,
        check=True,
    )
(venv_dir / ".installed").touch()
```

Furthermore, for the tempdir, IMHO we should use `tempfile.TemporaryDirectory` and the context manager it provides:

```python
with tempfile.TemporaryDirectory() as tmpdir:
    # do stuff with tmpdir
# tmpdir is  automatically deleted at this point
```

but here we need the tempdir to live for as long as `resolve_deps()` is running. I think we can solve this by stacking context managers. In other words, we start by transforming this function into a context manager:
```python
from contextlib import contextmanager
import venv
...
@contextmanager
def temp_venv(requirements: Iterable[str]) -> Iterator[Path]:
    """Provide a context with a temporary venv where the given requirements are installed."""
    with tempfile.TemporaryDirectory() as tmpdir:
        venv_dir = Path(tmpdir)
        venv.create(...) # and the other stuff above
        yield venv_dir
```
This can now be used from another function like this:
```python
@contextmanager
def pypi_deps(requirements: Iterable[str]) -> Iterator[LocalPackageLookup]:
    with temp_venv(requiremnts) as venv_dir:
        yield LocalPackageLookup(venv_dir)
```
And from there we can finally use this inside `resolve_dependencies()`:

```python
    if pull_deps:
        context = pypi_deps(deps)
    else:
        context = contextlib.nullcontext(LocalPackageLookup(pyenv_path))

    with context as local_packages:
        for name in deps:
            # same loop body as before
    # no need to call shutil.rmtree() the context managers will auto-unwind and the tempdir will be deleted.
```

(I like `@contextmanager` almost as much as I like `yield`... 😻)

Reply:
I added a single context manager to setup and tear down the environment in revised [a41e0e1](https://github.com/tweag/FawltyDeps/pull/252/commits/a41e0e11896567bf0803c4e8614de8b66346209a).